### DISCUSSION TITLE
How to create a role at application level?

### DISCUSSION THREAD

**Discussion by abdonrd** [directus/directus#11790] [Archive]
When we create a Directus application, we already have two roles: `Administrator` & `Public`.

My application is going to require another role to be able to be executed.
Is there a way to create that role in a bootstrap step or something similar?

As the `Public` role, it would also be a role that could never be deleted.

The use case is that the application programmatically assigns that role to users who meet certain conditions.

**Comment by azrikahar** [comment]
There isn't technically a "seed on/after bootstrap" step as of now (in case that's what you have in mind, but this is indeed discussed before!), but you should be able to achieve something similar via one of the alternatives:

- Create a [Custom Migration](https://docs.directus.io/extensions/migrations/#custom-migrations) to insert the role, as migrations will be ran on bootstrap by default as well.
- Create a custom script and create the role via the [Roles](https://docs.directus.io/reference/system/roles/) endpoint (or using the SDK).

**Reply by abdonrd** [reply]
Ok! Thanks for the info!

Trying the custom migration, now I have this:

```javascript
module.exports = {
  async up(knex) {
    const role = {
      name: 'Role name',
      icon: 'supervised_user_circle',
      enforce_tfa: false,
      admin_access: false,
      app_access: false,
    };

    await knex('directus_roles').insert(role);

    console.log(`The ${role.name} has been added.`);
  },
};
```

But then I have:

> 🚨 error insert into "directus_roles" ("admin_access", "app_access", "enforce_tfa", "icon", "name") values ($1, $2, $3, $4, $5) - null value in column "id" of relation "directus_roles" violates not-null constraint

To fix it I can manually add the `id` to the `role` object. It's expected, right?
Or should the `directus_role` auto-generate the `id`?

And if I need to add it manually, there is any helper to create the `uuid`?

**Reply by azrikahar** [reply]
Oh ya that's expected. You can either use the [uuid package](https://www.npmjs.com/package/uuid) ([the entry in Directus api package.json](https://github.com/directus/directus/blob/2938821be05eaf195872c34eed709ac9b4a430b4/api/package.json#L158)) or even better the [crypto.randomUUID()](https://nodejs.org/ca/blog/release/v14.17.0/#uuid-support-in-the-crypto-module) ([documentation](https://nodejs.org/docs/latest-v14.x/api/crypto.html#crypto_crypto_randomuuid_options)) function provided you're on Node v14.17+ 👍

**Reply by abdonrd** [reply]
Ok! Thanks again! 👏

**Reply by abdonrd** [reply]
Ouch! I get with an issue related to this: https://github.com/directus/directus/discussions/11798
